keps/sig-etcd/4647-cluster-feature-gate/README.md

```diff
@@ -0,0 +1,482 @@
+
+Now the developer could just send a client request to etcd cluster to query if a feature is enabled on the whole cluster. 
+
+#### Story 3
+
+In a HA cluster, users would like to turn off an enabled feature. They need to restart each server node one by one with feature changed from enabled to disabled. Today, after the restarting process begins and before all nodes are restarted, the behavior of the feature is really undefined: it is enabled on some nodes and disabled on the other nodes.
+
+Now with the cluster feature gate, we can deterministically tell if the feature is enabled or disabled for the whole cluster at any moment.
+
+This story applies similar to the upgrade/downgrade scenarios as well.
+
+### Notes/Constraints/Caveats (Optional)
+
+#### Could it done on the client side instead of inside etcd? 
+One could argue that it is easy to know if a feature enabled for the whole cluster by querying each member if the feature is enabled on the client side. But there are several caveates with this approach:
+* not all members are available all the time. if some member is not available, it is questionable if the feature can be used, and that could break the HA assumption. 
+* some feature might change how a raft message is sent and applied, and the order of messages relative to the index the feature is enabled is critical. Here the consistent index is more important than the time, and it is hard to get that outside etcd. 
+
+#### Should we prevent a new member with an inconsistent feature gate from joining the cluster?
+
+One way to prevent inconsistencies of feature gate between different members is to prevent a new member with an inconsistent feature gate from joining the cluster in the first place. 
+We do not think this is the right way because we want to keep the ability to turn on/off a feature in a rolling sequence without bringing down the whole cluster. 
+
+
+### Risks and Mitigations
+
+#### Backport Risks
+
+Because we are adding a new type of raft request, if we want to support feature gate as soon as in 3.6, we have to backport the proto changes and add no-op apply logic in 3.5. 
+There is the potential risk of changing a stable release. But we do not think this is a real risk, because it should be a no-op change in 3.5. 
+When upgrading etcd cluster from 3.5 to 3.6, the users would need to upgrade to the latest 3.5 release first.
+
+If we do not backport to 3.5, we would need to add no-op changes in 3.6, and wait till 3.7 to add the real feature gate capabilities, which could be a long cycle of several years.
```

Comment:
Don't understand this section. Can you clarify what backport you mean? Of the proposed mechanism or a feature gate.